api/plugin/vproto/api.proto

```diff
@@ -0,0 +1,317 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+syntax = "proto3";
+
+import "github.com/gogo/protobuf/gogoproto/gogo.proto";
+import "google/protobuf/empty.proto";
+import "k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto";
+
+package vproto;
+
+option go_package = "github.com/containerd/nri/api/plugin/vproto;vproto";
+
+option (gogoproto.goproto_unkeyed_all) = false;
+option (gogoproto.goproto_unrecognized_all) = false;
+option (gogoproto.goproto_sizecache_all) = false;
+option (gogoproto.goproto_getters_all) = true;
+
+
+//
+// Plugin is the ttRPC service than NRI plugins must implement.
+//
+service Plugin {
+  // Configure the plugin.
+  //
+  // Static plugins, IOW plugins that are started by NRI, get passed any
+  // plugin-specific raw configuration data found in the NRI configuration.
+  // Dynamic plugins, IOW plugins started by means external to NRI that
+  // register themselves by connecting to the well-known NRI socket, get
+  // empty configuration data.
+  //
+  // A plugin responds to a configuration request with the set of events it
+  // wants to subscribe to.
+  //
+  rpc Configure(ConfigureRequest) returns (ConfigureResponse);
```

Comment:
We can do that, but 

> How a plugin communicates errors like bad plugin configuration or required hardware features missing from the environment?
> 
> As a user I would like to see error messages that affect plugin life cycle (like "configuration syntax error, plugin disabled", "invalid resource management policy 'pofpools', plugin disabled") in the container runtime log. It would be nice to see other important messages like "PMEM not present, cold start disabled", too.
> 
> Do you think we could extend at least `ConfigureResponse` and `SynchronizeResponse` with some sort of status message fields?

Yes, that is of course possible and it might be necessary. But how would that be helpful in these cases ? I think we need to carefully consider whether and under what circumstances we want to allow errors to be returned at two different levels: within the protocol and at the transport (ttRPC) level. 

Just looking at the ability to propagate errors back from the plugin to the runtime, you can already return errors extra-protocol, at the ttRPC level, to any of the requests. All such errors should already be logged by the NRI prototype client and should show up among the stream of log messages produced by the runtime.

Looking at the method of reporting errors in-band protocol vs. extra-protocol, I think all errors reported extra-protocol should be considered fatal and they should stop any further transactions/processing which would have require the failed operation to succeed instead. This is also how the prototype currently handles errors returned by the ttRPC calls. And I think that is also the right course of action for both of the above-mentioned failures.

A failed `Configure()` request means that the plugin could not transition itself to an operational state with the configuration data it was provided by NRI (or which it has acquired by other means). The only reasonable course of action I can see for the runtime is to shut down the plugin and close its connection. For a statically configured plugin (already available and started when the runtime starts) I tend to think that it would also make sense to prevent the runtime from starting up as well...

Similarly a failed `Synchronize()` request means that, given the set of already running set of pods and containers, the plugin failed to find a set of container updates which would result in a running state considered valid by the plugin. Again, I think the only reasonable course of action for the runtime is to shut down the plugin and close its connection.

If we decide to provide two alternative channels for errors, IMO there should be some clear advantages and (semantic) rules as to which errors to propagate via which channel. Also there should be some clear semantic distinction between how NRI/the runtime will react to errors sent at the different channels. After all, if the reaction to all of them is the same, it is difficult to justify why there are two ways to send them back.